File: 0087-scramble-string/0087-scramble-string.js
/**
 * @param {string} s1
 * @param {string} s2
 * @return {boolean}
 */
var isScramble = function (s1, s2) {
  let dp = new Map();

  function isScrambleHelper(s1, s2) {
    if (s1 === s2) {
      return true;
    }
    if (dp.has(s1 + s2)) {
      return dp.get(s1 + s2);
    }

    if (s1.split("").sort().join("") !== s2.split("").sort().join("")) {
      dp.set(s1 + s2, false);
      return false;
    }

    for (let i = 1; i < s1.length; i++) {
      if (
        (isScrambleHelper(s1.substring(0, i), s2.substring(0, i)) &&
          isScrambleHelper(s1.substring(i), s2.substring(i))) ||
        (isScrambleHelper(s1.substring(0, i), s2.substring(s1.length - i)) &&
          isScrambleHelper(s1.substring(i), s2.substring(0, s1.length - i)))
      ) {
        dp.set(s1 + s2, true);
        return true;
      }
    }
    dp.set(s1 + s2, false);
    return false;
  }

  return isScrambleHelper(s1, s2);
};

File: 0087-scramble-string/0087-scramble-string.rb
# @param {String} s1
# @param {String} s2
# @return {Boolean}
def is_scramble(s1, s2)
    dp = {}
    
    def is_scramble_helper(s1, s2, dp)
        return true if s1 == s2
        return dp[[s1, s2]] if dp.include?([s1, s2])
        
        return false if s1.chars.sort.join != s2.chars.sort.join
        
        res = false
        for i in 1...s1.length
            if (is_scramble_helper(s1[0...i], s2[0...i], dp) && is_scramble_helper(s1[i...s1.length], s2[i...s2.length], dp)) || 
               (is_scramble_helper(s1[0...i], s2[s1.length - i...s2.length], dp) && is_scramble_helper(s1[i...s1.length], s2[0...s1.length - i], dp))
                res = true
                break
            end
        end
        
        dp[[s1, s2]] = res
        return res
    end
    
    is_scramble_helper(s1, s2, dp)
end